### DISCUSSION TITLE
Startup time

### DISCUSSION THREAD

**Discussion by kybos** [muse-sequencer/muse#1153] [Team Posts]
Please don't stone me if the idea is a complete nonsense... but the start of MusE takes quite long (with default options), about 5 seconds on a high-end machine with 16 cores. Compared with QTractor with less than 0.5 secs... Clearly the initialization of the devices, plugins etc. takes its time, but could not that be done in separate thread(s)? I assume that the availability of the devices is checked by MusE anyway before they are really used...

**Comment by terminator356** [comment]
Is that before or after a plugin cache update has been done?
Upon startup, a cache update takes a long time but after that, subsequent restarts of MusE should
 be as fast as they can be (currently).
Are you suffering from repeated slow cache updates on every restart? Let me know.

Apart from that, I was also thinking the same thing the other day about offloading a few things to threads.
I can't remember what, not sure if it was startup time...

**Comment by kybos** [comment]
It's the same on each start, otherwise I would not mention it ;-).
I will check in more detail if you think it should not take so long.
I guess it is related to the plugins. If I use the switches to skip the loading of all plugin types, then it's considerably faster (but still far from QTractor ;-) ).

**Comment by terminator356** [comment]
I think QTractor does not scan plugins or even fill a plugin list until you actually open its plugin dialog.
When it opens you can see the time it takes to scan. Or if you hit the rescan button.
That would be the equivalent time that MusE takes to fill plugin lists and so on at startup.
We do that so plugins are readily available from the start and can be loaded from a song file.
Conversely, I think when QTractor loads a song it just loads the related plugins rather than filling
 a whole list to pick from beforehand like we do, although it does quickly load its cache for that.

One definite improvement in our long plugin cache scanning, when it happens,
 would be multiple scan threads. Currently I use just one thread to check a plugin
 and wait till it's done before moving on to the next plugin.
Instead I should farm out several threads at once each scanning a different plugin.

**Comment by spamatica** [comment]
Isn't QTractor sort of the exception? It's insanely fast at starting...
Did some recordings with a friend that uses Cubase a year ago and I was
actually kind of surprised with how long it took to start, going through
plugins like it was perfectly normal. I would have assumed that would have
been improved by now (though I'm not sure how old that version was).

We can of course aspire to be as fast as QTractor, waiting is annoying.
Though with the cache we've already improved the startup time quite a bit
(and stabilized startup hugely, no more crashes!).
While adding the track menu I came across a comment which hinted that maybe
it should be done in a thread. There are probably a lot of similar stuff
that can be parallellized.

There is still something weird with the cache though I haven't managed to
get to the bottom of it, my recording computer (which has a lot of plugins)
still seems to always refresh the cache. Will get back to you guys on that.

I have a question though regarding LV2, is that handled differently from a
cache perspective? Reading the code (and what is output in the terminal) it
looks like it always goes through the LV2 plugins?


Den fre 14 feb. 2020 kl 00:25 skrev Tim <notifications@github.com>:

**Comment by terminator356** [comment]
Yes. Until last week, I created a cache file for LV2, along with the others, just to 'be thorough'.
It turned out to be useless, unnecessary, and caused trouble.
LV2 uses text description files, so scanning and caching is useless.
So, actually both cache scanning time **and** normal startup time have been improved
 since there's no LV2 cache file to either create, or check at startup.

**Comment by spamatica** [comment]
Alright. I suppose that it might be marginally faster to read from the cache, we'll see if there are some drawbacks.

Regarding startup speed in general, let me just state that if anyone @kybos? wants to pursue paralleliize the startup - if there is a risk it will affect stability, I'm against. ;)

I'm quite happy with how stable MusE is, at least for my workflow, crashes do occur but they are far between.
Qtractor on the other hand - I don't start it often but when I do (usually to try if some plugin works there that don't work in MusE) I often end up in triggering something that crashes the application. 
I do realize the comparison might be unfair as I usually don't know what I'm doing in Qtractor and similar clicking-around in MusE might be equally bad...  but anyway ;)

**Comment by terminator356** [comment]
Should be no drawbacks, only benefits. Simpler code, no LV2 cache file to worry about,
 faster startup time, less code headaches.
Look at it this way: The only reason we need a cache is for all those other plugin types
 that have only an *.so file to (dangerously) open and discover.
If they all worked like LV2's text-based discovery system, we wouldn't need caches ;-)
I'm eager to see how VST3 works and support it. It's open source now, you know.
I played around with the SDK a few years back. It seems your beloved A.D. is not yet VST3?

I'm always impressed when I have to wade through QTractor's code.
Clean, modern, easy to navigate, familiar Qt territory.
There are things done superbly well in there.
But there are a few things that we do more superbly.  [ Cue the trumpets. ]

About startup time: I must point out that there is a (slight? not sure) bottleneck in
 that I actually use two plugin systems. It's hard to explain why, you'd have to read the
 ChangeLog to see why, but my ultimate goal is to eliminate our existing rather messy
 plugin system with the newer one I wrote.
The bottleneck is that I read all the caches into this newer plugin system,
 and then fill up our existing plugin system from that newer system.
I want to eliminate that but obviously it will take big effort to finally replace our existing system.
So this was the best I could do for now until that goal can be reached.

And there's actually a third plugin system involved that I'm trying to merge with all of this:
That's the special plugin system in our Deicsonze and SimpleDrums MESS.
Again, for now, until some goals can be achieved, I fill up both our existing plugin system
 and Deicsonze/SimpleDrums's from this newer system. Ah, interface layers...

**Comment by kybos** [comment]
> Regarding startup speed in general, let me just state that if anyone @kybos? wants to pursue paralleliize the startup - if there is a risk it will affect stability, I'm against. ;)

I had not intention to do that, it was just an idea, seeing that Tim is improving this area anyway. So I rely on him to do the right thing, sure I'm not going to start modifying the same code in parallel (especially after reading about the three different plugin systems ;-)...).
Generally I don't share your skepticism. Multithreading is nothing esoteric any more, there is a very good support for it in Qt5. If done correctly, there should be no more risk than from any other change. Nowadays when even the weakest PC has a dual or quad core processor, it's a good idea to adapt the programming paradigm to use the available resources where it makes sense.

> I'm quite happy with how stable MusE is, at least for my workflow, crashes do occur but they are far between.
> Qtractor on the other hand - I don't start it often but when I do (usually to try if some plugin works there that don't work in MusE) I often end up in triggering something that crashes the application.
> I do realize the comparison might be unfair as I usually don't know what I'm doing in Qtractor and similar clicking-around in MusE might be equally bad... but anyway ;)

I have the same experience with the (in)stability of QTractor - that was the reason I started using MusE, after all...

**Comment by spamatica** [comment]
> Generally I don't share your skepticism. Multithreading is nothing esoteric any more, there is a very good support for it in Qt5. If done correctly, there should be no more risk than from any other change. Nowadays when even the weakest PC has a dual or quad core processor, it's a good idea to adapt the programming paradigm to use the available resources where it makes sense.
> 
The older I get, the more skeptic I get ;)
Probably lots of things can be done safely, like the one I mentioned, building menu content. 
You can probably safely ignore my ranting :D
